src/main/java/bio/terra/service/filedata/google/firestore/FireStoreDependencyDao.java

```diff
@@ -128,49 +130,57 @@ private void batchStoreSnapshotFileDependencies(
             getFutures.add(query.get());
         }
 
-        try {
-            // Scan the lookup results and launch the sets in parallel
-            int index = 0;
-            for (ApiFuture<QuerySnapshot> future : getFutures) {
-                String fileId = batch.get(index);
-
-                List<QueryDocumentSnapshot> documents = future.get().getDocuments();
-                switch (documents.size()) {
-                    case 0: {
-                        // no dependency yet. Let's make one
-                        FireStoreDependency fireStoreDependency = new FireStoreDependency()
-                            .snapshotId(snapshotId)
-                            .fileId(fileId)
-                            .refCount(1L);
-                        DocumentReference docRef = depColl.document();
-                        setFutures.add(docRef.set(fireStoreDependency));
-                        break;
-                    }
-
-                    case 1: {
-                        // existing dependency; increment the reference count
-                        QueryDocumentSnapshot docSnap = documents.get(0);
-                        FireStoreDependency fireStoreDependency = docSnap.toObject(FireStoreDependency.class);
-                        fireStoreDependency.refCount(fireStoreDependency.getRefCount() + 1);
-                        setFutures.add(docSnap.getReference().set(fireStoreDependency));
-                        break;
+        for (int i = 0; i < RETRIES; i++) {
```

Comment:
Maybe I'm missing something, but I don't see how this loop exits early if no errors occur. Or is it OK to always run the body three times, even if no errors occur?

In this case where an error does occur, and the loop retries the body, is there no work needed to remove any futures that executed successfully and don't need to be retried?